From f5de7e2c7b18b69ade3f07bab5cc1ac73e63a57d Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Wed, 08 Jun 2022 15:01:33 +0300
Subject: [PATCH] [SW-70527] gaudi2: ras: sm, rot sram rings repair

This patch added support for BIRA, BISR and BIST operations
for SM, ROT SRAM rings. Also after BIST SM, ROTATOR
unit reset is done to avoid ECC.

Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
Change-Id: I6a22fc3946d4ab3a2d584776965aeee2d755aea6
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index 8866c11..3a7e4c4 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -33,6 +33,8 @@
 #define DELAY_7P5K_USEC			7500
 #define DELAY_1K_USEC			1000
 #define DELAY_17_USEC			17
+#define DELAY_8K_USEC			8000
+#define DELAY_57_USEC			57
 
 #define GAUDI2_DFX_NUM			5
 #define GAUDI2_MAX_RING_NUM		28
@@ -272,6 +274,59 @@
 	}
 }
 
+static void gaudi2_sob_rot_ring_pattern(enum gaudi2_sram_self_repair_op op)
+{
+	uint32_t expected_ones, expected_zeros;
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL35, mmPSOC_DFT_APB_WORD0);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL2, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_8K_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL3, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_57_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL4, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_7P5K_USEC);
+	} else {
+		preboot_puts("Invalid sram repair operation\n");
+		return;
+	}
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL36, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000010100000010100000010100;
+	expected_zeros = 0b00000000001011100001011100001011;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST)
+		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
+	else
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000000000000000001010000000;
+	expected_zeros = 0b00000000000000000010000101110000;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL35, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+
+		regmap_writel(GAUDI2_RING_PATTERN_VAL37, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+		expected_ones = 0b00000000000000000000000000000000;
+		expected_zeros = 0b00000000000000000010010010010000;
+		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	}
+}
+
 static void gaudi2_sram_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
@@ -354,7 +409,7 @@
 
 static void (*gaudi2_sram_region_ring_pattern[])(enum gaudi2_sram_self_repair_op op)  = {
 				[SRAM_RING_TYPE_TPC] = gaudi2_tpc_ring_pattern,
-				[SRAM_RING_TYPE_SOB_ROT] = NULL,
+				[SRAM_RING_TYPE_SOB_ROT] = gaudi2_sob_rot_ring_pattern,
 				[SRAM_RING_TYPE_SRAM] = gaudi2_sram_ring_pattern,
 				[SRAM_RING_TYPE_XBAR_MID] = NULL,
 				[SRAM_RING_TYPE_VDEC] = NULL,
@@ -378,7 +433,8 @@
 		for (ring = 0; ring < ring_count_table[dfx]; ring++) {
 			ring_type = gauid2_ring_type_table[dfx][ring];
 			if (ring_type == SRAM_RING_TYPE_SRAM ||
-				ring_type == SRAM_RING_TYPE_TPC) {
+				ring_type == SRAM_RING_TYPE_TPC ||
+				ring_type == SRAM_RING_TYPE_SOB_ROT) {
 				gaudi2_sram_id_sel();
 				gaudi2_sram_ring_sel(dfx_value_table[dfx], ring + 1);
 				gaudi2_sram_region_ring_pattern[ring_type](op);
@@ -496,7 +552,7 @@
 /* TODO: initialize internal tpc srams which
  * are not connected to reset block
  */
-static void gaudi2_reset_tpc(void)
+static void gaudi2_reset_engines(void)
 {
 	u32 tpc_iso_val;
 
@@ -505,11 +561,15 @@
 
 	regmap_writel(PSOC_RESET_CONF_TPC_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_TPC_UNIT_RST);
 	regmap_writel(PSOC_RESET_CONF_TPC_DIV_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_TPC_DIV_UNIT_RST);
+	regmap_writel(PSOC_RESET_CONF_ROTATOR_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_ROTATOR_UNIT_RST);
+	regmap_writel(PSOC_RESET_CONF_SM_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_SM_UNIT_RST);
 	regmap_writel(PSOC_RESET_CONF_UNIT_RST_N_IND_MASK, mmPSOC_RESET_CONF_UNIT_RST_N);
 	udelay(DELAY_1K_USEC); /* TODO: confirm the delay */
 	regmap_writel(0, mmPSOC_RESET_CONF_UNIT_RST_N);
 	regmap_writel(0, mmPSOC_RESET_CONF_TPC_UNIT_RST);
 	regmap_writel(0, mmPSOC_RESET_CONF_TPC_DIV_UNIT_RST);
+	regmap_writel(0, mmPSOC_RESET_CONF_SM_UNIT_RST);
+	regmap_writel(0, mmPSOC_RESET_CONF_ROTATOR_UNIT_RST);
 	udelay(DELAY_1K_USEC); /* TODO: confirm the delay */
 
 	/* Restore original iso value */
@@ -525,6 +585,6 @@
 	/* Get back to normal from DFT mode */
 	regmap_writel(DFT_APB_CONTROL_DEFAULT_VAL, mmPSOC_GLOBAL_CONF_DFT_APB_CONTROL);
 	gaudi2_scrub_sram(); /* TODO: Replace this by UNIT reset (Refer SIV-1088) */
-	gaudi2_reset_tpc();
+	gaudi2_reset_engines();
 	preboot_puts("sram bist done\n");
 }
diff --git a/arch/gaudi2/common/common.h b/arch/gaudi2/common/common.h
index 74d2e23..fd5ad8a 100644
--- a/arch/gaudi2/common/common.h
+++ b/arch/gaudi2/common/common.h
@@ -257,6 +257,9 @@
 #define GAUDI2_RING_PATTERN_VAL32		0x11111
 #define GAUDI2_RING_PATTERN_VAL33		0x100007f
 #define GAUDI2_RING_PATTERN_VAL34		0x1000031
+#define GAUDI2_RING_PATTERN_VAL35		0x100002a
+#define GAUDI2_RING_PATTERN_VAL36		0x100002e
+#define GAUDI2_RING_PATTERN_VAL37		0x1000016
 
 
 void gaudi2_axi_drain_ena(void);
